logics/impls/controller/mod.rs

```diff
@@ -718,27 +694,239 @@ impl<T: Storage<Data>> Internal for T {
         pool: AccountId,
         redeemer: AccountId,
         redeem_amount: Balance,
-        pool_attribute: Option<PoolAttributes>,
+        pool_attributes: Option<PoolAttributesForWithdrawValidation>,
     ) -> Result<()> {
-        let caller_pool = if let Some(_pool_attribute) = pool_attribute {
-            Some((pool, _pool_attribute))
-        } else {
-            None
-        };
-        let (_, shortfall) = self._get_hypothetical_account_liquidity(
-            redeemer,
-            Some(pool),
-            redeem_amount,
-            0,
-            caller_pool,
-        )?;
-        if !shortfall.is_zero() {
-            return Err(Error::InsufficientLiquidity)
-        }
+        if let Some(oracle) = self._oracle() {
+            let mut asset_price: u128 = 0;
+            let mut liquidation_threshold: u128 = 0;
+            let (shortfall, account_data) = {
```

Comment:
(Although either way of assigning the result of a calculation is fine.) Since we are using two different way, let's align them with one of them.

```rust
let (shortfall, account_data, asset_price, liquidation_threshold) = { ... }
```